Consolidate Legato AF integration

splasky opened the pull request:
These commits do the following things:
1. Fix warning during compiling.
2. Fix snprintf format on different platform.
3. Consolidate Legato AF intergration.

close #20

jserv commented:
Use precise wording. Tweak the git commit messages accordingly.

howjmay commented on Makefile:
```diff
@@ -51,27 +52,36 @@ mbedtls_make:
 	$(MAKE) -C $(UTILS_PATH)
 test: $(TEST_PATH) $(UTILS_OBJS)
 	$(MAKE) -C $(TEST_PATH)
+ta_app: $(TALIB)
```
May you change it to `ta_leagto_app` here, since I think we should state clearly that this rule would generate legato app only.

howjmay commented on utils/utils.h:
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (C) 2018-2019 BiiLabs Co., Ltd. and Contributors
```
2019-2020

howjmay commented on utils/utils.c:
```diff
@@ -0,0 +1,75 @@
+
+static char addr_log_template[] = "\n%s\n";
+
+void gen_trytes(uint16_t len, char *out) {
```
May you help me to rename it as `gen_rand_trytes()`

howjmay commented on utils/crypto_utils.h:
```diff
@@ -19,12 +19,12 @@ extern "C" {
+int ta_encrypt(unsigned char *plaintext, int plaintext_len, unsigned char *ciphertext, int ciphertext_len,
+               uint8_t iv[16], uint8_t key[AES_BLOCK_SIZE * 2], uint8_t device_id[IMSI_LEN + 1]);
+int ta_decrypt(unsigned char *ciphertext, int ciphertext_len, unsigned char *plaintext, int plaintext_len,
+               uint8_t iv[16], uint8_t key[AES_BLOCK_SIZE * 2]);
```
May you assign the size of IV with macro `AES_IV_SIZE` and define one more macro for `AES_KEY_SIZE`?
Btw are we using aes cbc here? If we are using cbc, then I think it is better to name the macro of iv size as `CBC_IV_SIZE`

howjmay commented on endpointComponent/endpoint.c:
```diff
@@ -0,0 +1,96 @@
+      // Init http session. verify: check the server CA cert.
+      char msg_body[1024];
+      gen_tryte_msg(tryte_msg, addr, msg_body);
+      if (send_https_msg(HOST, PORT, API, msg_body, 1024) != HTTP_OK) {
```
We should allow user to use other `host` and `port` here

splasky replied:
Enter by the command line or config file?

howjmay replied:
How about using cli at the first step? 
I think it may be easier to implement

howjmay replied:
btw, Where is `HTTP_OK` defined?

splasky replied:
Already defined inside defined_error.h.

howjmay replied:
thx

jserv commented on Makefile:
```diff
@@ -29,12 +29,13 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
 
-OBJS = main.o $(HTTP_PARSER_PATH)/http_parser.o $(UTILS_OBJS) $(CONNECTIVITY_OBJS)
+OBJS = main.o http_parser.o $(UTILS_OBJS) $(CONNECTIVITY_OBJS)
+TALIB = libta_endpoint.so
```
Change it to `TA_LIB`.

jserv commented on utils/utils.h:
```diff
@@ -0,0 +1,66 @@
+ */
+int send_https_msg(char const *const host, char const *const port, char const *const api, const char *msg,
+                   const int msg_len);
+#endif /* end of include guard: UTILS_H_JKZACOAY */
```
What is `UTILS_H_JKZACOAY`?

splasky replied:
This is automatically generate by vim snippets.

howjmay commented on endpointComponent/endpoint.c:
```diff
@@ -0,0 +1,96 @@
+      uint8_t id[IMSI_LEN + 1] = {0};
+
+      if (get_aes_key(private_key) != 0) {
+        LE_ERROR("%s\n", "get aes key error");
```
Are we able to use a macro to wrapper `LOG_ERROR` to wrap `LE_ERROR`, since we may need to run ta-endpoint on other platform (i.e. testing on linux on laptop).
Maybe we could open another issue for this task

splasky replied:
Create a new issue for these question?
Now we have lot of DEBUG or some macro flags during compiling for different platform and test target.

splasky replied:
> Are we able to use a macro to wrapper `LOG_ERROR` to wrap `LE_ERROR`, since we may need to run ta-endpoint on other platform (i.e. testing on linux on laptop).
> Maybe we could open another issue for this task

Create new issue at #28

howjmay commented on utils/utils.c:
```diff
@@ -0,0 +1,75 @@
+  snprintf(addr_log, 83, addr_log_template, next_addr);
+  fputs(addr_log, fp);
+  fclose(fp);
+  return 0;
```
We should use some status code declared in enum instead of int here.

howjmay commented on utils/crypto_utils.c:
```diff
@@ -134,16 +135,16 @@ int aes_decrypt(const unsigned char *ciphertext, int ciphertext_len, const unsig
-    n = (ciphertext_len - i > AES_BLOCK_SIZE) ? 16 : (int)(ciphertext_len - i);
+  for (int i = 0; i < ciphertext_len; i += CBC_IV_SIZE) {
+    memset(buf, 0, CBC_IV_SIZE);
+    n = (ciphertext_len - i > CBC_IV_SIZE) ? 16 : (int)(ciphertext_len - i);
```
Do we need to change the `16` here into `CBC_IV_SIZE`?

howjmay commented on utils/crypto_utils.c:
```diff
@@ -93,16 +94,16 @@ int aes_encrypt(const unsigned char *plaintext, int plaintext_len, const unsigne
-    n = (new_len - i > AES_BLOCK_SIZE) ? 16 : (int)(new_len - i);
+  for (int i = 0; i < new_len; i += CBC_IV_SIZE) {
+    memset(buf, 0, CBC_IV_SIZE);
+    n = (new_len - i > CBC_IV_SIZE) ? 16 : (int)(new_len - i);
```
ditto

jserv commented on endpointComponent/endpoint.c:
```diff
@@ -0,0 +1,142 @@
+static int PORT = TA_PORT;
+static char HOST[1024] = TA_HOST;
+
+//--------------------------------------------------------------------------------------------------
```
Remove this line.

jserv commented on endpointComponent/endpoint.c:
```diff
@@ -0,0 +1,142 @@
+ * Prints the on-line help and exits.  This function is called when "-h" or "--help" appears on
+ * the command-line or when the help command is invoked.
+ **/
+//--------------------------------------------------------------------------------------------------
```
Remove this line.

jserv commented on utils/crypto_utils.c:
```diff
@@ -29,7 +30,7 @@ int get_device_id(const char *device_id) {
   }
 
-  strncpy((char *)device_id, imsi, IMSI_LEN);
+  strcpy((char *)device_id, imsi);
```
Why did you invoke `strcpy`?

splasky replied:
After gcc version > 8 `strncpy` is for different usage. See `-Wstringop-truncation`

jserv replied:
We need security advice. Can you show some reference materials?

jserv commented on utils/uart_utils.c:
```diff
@@ -81,11 +81,13 @@ char *uart_read(const int fd) {
   if (rdlen > 0) {
-    // printf("buf = %s\n", buf);
+#ifdef DEBUG
+    printf("buf = %s\n", buf);
```
Is it necessary?

splasky replied:
I think all of macro flags during compiling should be reduce. But I will not solve in this pull request. Should be do in refactor stage.

jserv replied:
Create a new issue for cleaning up existing codebase.

splasky replied:
New issue at #33

jserv commented on Makefile:
```diff
@@ -29,49 +29,67 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
 
-clean: clean_client clean_third_party clean_test
+## Build ta-endpoint legato app
+ta_legato_app:
```
The target name is misleading. I would expect `legato` simply.

jserv commented on Makefile:
```diff
@@ -29,49 +29,67 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
+## Build ta-endpoint legato app
+ta_legato_app:
+	# TODO: notify should make ta_lib with arm compiler first
+	mkapp -v -t wp77xx ta-endpoint.adef
```
You shall detect Legato AF in advance.

splasky replied:
I have already tried this. But how can we detect current bash setting like after command `source legat_bashrc` environment inside makefile?

jserv replied:
It sounds good. But, you shall improve the build instructions.

jserv commented:
Use `git rebase -i` to rework existing git commits. I would like to functional changes rather than trivial patches.

howjmay commented on endpointComponent/Component.cdef:
```diff
@@ -0,0 +1,41 @@
+sources:
```
Could we put it under a folder for different hardware instead putting under root directory

splasky replied:
Related at #31

jserv commented on Makefile:
```diff
@@ -29,49 +29,66 @@ INCLUDES = -I$(THIRD_PARTY_PATH)/http-parser -I$(MBEDTLS_PATH)/include -I$(ROOT_
+
+## Prints help for targets with comments
+help:
+	@cat $(MAKEFILE_LIST)|awk '/^##.*$$/{l1=$$0;getline;l2=(l1 "##" $$0); print l2 $$0}' | awk -F"##" '{split($$3,t,":");printf "\033[36m%-30s\033[0m %s\n",t[1],$$2}'
```
Split into several pretty lines.

jserv commented on ta-endpoint.adef:
```diff
@@ -0,0 +1,12 @@
+executables:
+{
+    taEndpoint = (endpointComponent)
```
Why? `endpointComponent` sounds weird. Can you simply use `component(s)`?

splasky replied:
We will have other components inside directory. The enpointComponent just for endpoint's  component. Change to `components` will be misleading. How about change to `endpointLegato`?

splasky replied:
see also:https://github.com/legatoproject/legato-af/tree/master/apps/sample/mqtt

jserv replied:
No camelCase allowed for directory name.

jserv replied:
or simply use `legato`.

jserv commented on utils/utils.c:
```diff
@@ -0,0 +1,75 @@
+#include "utils.h"
```
Filename `utils/utils.c` is confusing. Can you use some functional names such `utils/url-rewriter.c`?

jserv commented on utils/url_msg.h:
```diff
@@ -0,0 +1,69 @@
+ * "LICENSE" at the root of this distribution.
+ */
+#ifndef URL_MSG_H
+#define URL_MSG_H
```
`url_msg.h` sounds weird. Change it to `url.h`.

jserv replied:
Also, add brief descriptions for this file.

jserv commented on utils/url_msg.h:
```diff
@@ -0,0 +1,69 @@
+#define URL_MSG_H
+
+#include <stdint.h>
+#define TA_HOST "tangle-accel.puyuma.org"
```
Don't hardcode hostname here.

jserv commented on legato/endpoint.c:
```diff
@@ -0,0 +1,154 @@
+static void PrintHelp(void) {
+  puts(
+      "NAME\n"
+      "        ta-enpoint - Transcation information with tangle accerlator.\n"
```
Always mention "Tangle-accelerator`.

jserv commented on legato/endpoint.c:
```diff
@@ -0,0 +1,154 @@
+  "POWEREDBYTANGLEACCELERATOR999999999999999999999999999999999999999999999999" \
+  "999999A"
+#define ADDR_LEN 81
+#define ADDR_LOG_PATH "addr_log.log"
```
The path is confusing. You can rename it as `ta-endpoint.log`.

jserv commented on legato/endpoint.c:
```diff
@@ -0,0 +1,154 @@
+
+#define MSG "%s:%s"
+
+#define TA_HOST "tangle-accel.puyuma.org"
```
Don't hardcoded hostname.

jserv commented on main.c:
```diff
@@ -1,95 +1,37 @@
+#define MSG "%s:THISISMSG9THISISMSG9THISISMSG"
+#endif
+
+#define TA_HOST "tangle-accel.puyuma.org"
```
Don't hardcoded hostname.

jserv replied:
You can set the default host via build system + macro.

jserv commented on utils/url.h:
```diff
@@ -0,0 +1,65 @@
+ * "LICENSE" at the root of this distribution.
+ */
+#ifndef URL_MSG_H
+#define URL_MSG_H
```
Should be `UTILS_URL_H`.

jserv commented on legato/Component.cdef:
```diff
@@ -0,0 +1,42 @@
+sources:
```
Use lower case `component.cdef`.

splasky replied:
This is the default build name for legato build system.

jserv commented on utils/url.c:
```diff
@@ -0,0 +1,77 @@
+  return ret;
+}
+
+int log_address(const char *path, uint8_t *next_addr, size_t addr_len) {
```
`log_address` is a misleading name. Ask @HowJMay for advice.

jserv commented:
Update top-level `README.md` accordingly.

howjmay reviewed (changes requested):
Add `close #20` in both git message and PR message

jserv commented:
Improve the git commit messages. We expect expressive wording.

jserv commented on utils/url.h:
```diff
@@ -0,0 +1,65 @@
+ * @param len length of output
+ * @param out output array
+ */
+void gen_rand_trytes(uint16_t len, uint8_t *out);
```
This is unrelated to URL. Move it into another header.

Commit:
[WIP]feat: Consolidate Legato AF intergration

Add legato build target into gitignore.

Add new file util/utils.h util/utils.c for utitlity functions.

Add build libta_endpoint.so library target for legato app ta_legato_app into Makefile.

Add ta_legato_app build target and cleanup target into Makefile.

Remove redefined macro inside `conn_http.h`.

Add new directory endpointComponent and files for legato app.

Move utility functions inside `main.c` into utils.h utils.c.

Add doxgen comment for utils.h.

Fix compile warning inside utils directory files.

Closes #20

Closes #29

howjmay commented:
We will integrate legato and ta-endpoint in another PR. The current one is too huge for taking future process, so I will close it.
